xcube/core/store/fs/impl/geotiff.py

```diff
@@ -58,27 +59,43 @@ def __init__(self,
         self._root = root
         self._path = data_id
         self._open_params = open_params
+        self._file_url = None
 
-    def _get_num_levels_lazily(self) -> int:
-        with rasterio.open(self._get_file_url()) as rio_dataset:
+    def _open_dataset_with_rasterio(self):
+        with rasterio.open(self._file_url) as rio_dataset:
             overviews = rio_dataset.overviews(1)
+        return overviews
+
+    def _get_num_levels_lazily(self) -> int:
+        self._file_url = self._get_file_url()
+        if isinstance(self._fs, s3fs.S3FileSystem):
+            fs: s3fs.S3FileSystem = self._fs
+            session = DatasetGeoTiffFsDataAccessor.create_env_session(fs)
```

Comment:
Redundant variable `session`. And I've seen the same code above. Avoid code duplication.